src/test/java/com/atlassian/db/replica/api/TestDualConnection.java

```diff
@@ -1382,4 +1401,21 @@ public void shouldKeepUsingReplicaAfterSettingTimeout() throws SQLException {
         assertThat(connectionProvider.getProvidedConnectionTypes())
             .containsOnly(REPLICA);
     }
+
+    @Test
+    public void shouldResetReadOnlyModeBeforeReleasingConnection() throws SQLException {
+        final ConnectionProviderMock connectionProvider = new ConnectionProviderMock();
+        final Connection connection = DualConnection.builder(
+            connectionProvider,
+            permanentConsistency().build()
+        ).build();
+        connection.setReadOnly(true);
+        connection.prepareStatement(SIMPLE_QUERY).executeQuery();
+        final Connection rawConnection = connectionProvider.singleProvidedConnection();
+        Mockito.reset(rawConnection);
+
+        connection.close();
+
+        Mockito.verify(rawConnection).setReadOnly(eq(false));
```

Comment:
How can a closed connection state affect anything? Can it be reopened?

Reply:
We are not the only ones adding magic between `jdbc` and business logic 😄 . connection pool intercepts connection#close and re-uses it instead of closing.

Reply:
Aaah 🤯 
Good point! Mental model updated 😉

Reply:
@dagguh How do you keep up with maintaining such complex model inside your head? 
Though its good we have those tests <3 
@wyrzyk such comment/code might be  valuable inside test for others